sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,338 @@
+import numpy as np
+
+from ..base import MetaEstimatorMixin, clone, BaseEstimator
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if not hasattr(estimator, "predict_proba"):
+        msg = "base_classifier ({}) should implement predict_proba!"
+        raise ValueError(msg.format(type(estimator).__name__))
+
+
+class SelfTrainingClassifier(MetaEstimatorMixin, BaseEstimator):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting pseudo-labels for the unlabeled data
+    and adding them to the training set.
+
+    The classifier will continue iterating until either max_iter is reached, or
+    no pseudo-labels were added to the training set in the previous iteration.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    selection_criterion : {'threshold', 'n_best'}, optional
+    (default='threshold')
+        The selection criterion used to select which labels to add to the
+        training set. If 'threshold', pseudo-labels with prediction
+        probabilities above `threshold` are added to the dataset. If 'n_best',
+        the `n_best` pseudo-labels with highest prediction probabilities are
+        added to the dataset.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold for use with `selection_criterion`='threshold'.
+        Should be in [0, 1).
+
+    n_best : int, optional (default=10)
```

Comment:
Should this be K best rather than n, for consistency with feature selection? It's not something we are generally very consistent about.